### ISSUE TITLE
Enhancement: Apply Filter automatically to messages outside Inbox

### ISSUE THREAD

**Issue by Pilkja** [RealRaven2000/FiltaQuilla#88]
I'm looking for a way to apply a filter action to messages as they are moved to a different folder. 
The standard "Apply filter when" unfortunately gives no option for a condition "when moved to folder". 

A Workaround could be using "rerun every 10 minutes", used on the arrival folder. But that a) doesn't seem to work (I tested it: "manually run" works, "rerun every 10 minutes" doesn't) and b) seems like quite some wasted cpu power.

Do you think this would be an interesting idea and is possible to implement or do you have any other idea for the problem?
Thanks!

**Comment by RealRaven2000** [comment]
It would be interesting, but we need some commitment from the Thunderbird Core developers to allow running filters automatically not just on the Inbox but on any folder. It's one of the architectural limitations of Thunderbird that all filters are stuck at the Inbox level at the moment, and that there is also no "global filter" mechanism. 

It would be great to have a special list of "global filters" with the additional condition "folder X", but I think we need to ask the Thunderbird Core team to implement this on their side.

**Comment by RealRaven2000** [comment]
related bug in quickFilters:

https://github.com/RealRaven2000/quickFilters/issues/30

**Comment by RealRaven2000** [comment]
Interesting article by Kent James (original author of FiltaQuilla), from an archive:

https://web.archive.org/web/20191006143156/http://www.mesquilla.net/mesquillacom/2010/04/12/what-filters-need/

**Comment by ThiloteE** [comment]
To put your links and your call for commitment by Thunderbird core developers into perspective:

Kent James rings a bell. When i was reading through Bugzilla a few months back, he seemed to have been one of the major filter experts, but is now inactive in that field. As far as i am aware, Thunderbird core (without extensions) has been stagnant in terms of filter development at least since ~ 5-7 years. So sad. It does not seem to be high on the priority list right now 😩 and if i look at this roadmap from 2010, it is sad to see that the interface for boolean logic has not been introduced, even though the underlying engine is able to do it since a whopping 12 years (!).

I know [Wayne Mery (:wsmwk)](https://bugzilla.mozilla.org/user_profile?user_id=29811) and [Magnus Melin [:mkmelin]](https://bugzilla.mozilla.org/user_profile?user_id=101158) regularly look at Bugzilla filter issues, but there is barely any work done to resolve these bugs. In 2021 alone, [7 filter issues](https://bugzilla.mozilla.org/buglist.cgi?order=Importance&product=Thunderbird&resolution=FIXED&bug_status=RESOLVED&query_format=advanced&component=Filters&classification=Client%20Software&classification=Developer%20Infrastructure&classification=Components&classification=Server%20Software&classification=Other) + [2 mail news core issues related to filters](https://bugzilla.mozilla.org/buglist.cgi?component=Filters&classification=Client%20Software&classification=Developer%20Infrastructure&classification=Components&classification=Server%20Software&classification=Other&resolution=FIXED&query_format=advanced&bug_status=RESOLVED&product=MailNews%20Core) were fixed and those were mostly regressions, not new features or enhancements and the years before that tell a similar story.

To my understanding, Thunderbird currently lacks the will, the expertise and workforce to tackle filter issues, because many filter developers are not active anymore. Not only that, there is technical debt. Many filter bugs stem from flaws in imap and pop3 protocoll or a wrong implementation of those standards (looking at you [filter loop issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1215933)) also, which blocks quite a few features, so implementation of a more flexible new message transfer protocol would be pretty high on the agenda, but maybe that is a different topic and may not necessarily block the introduction of applying filters automatically to messages outside Inbox.

Long story short: a rework of the filter interface and filter engine behind it is long overdue.

**Comment by RealRaven2000** [comment]
> To my understanding, Thunderbird currently lacks the will, the expertise and workforce to tackle filter issues, because many filter developers are not active anymore. Not only that, there is technical debt. Many filter bugs stem from flaws in imap and pop3 protocoll or a wrong implementation of those standards (looking at you [filter loop issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1215933)) also, which blocks quite a few features, so implementation of a more flexible new message transfer protocol would be pretty high on the agenda, but maybe that is a different topic and may not necessarily block the introduction of applying filters automatically to messages outside Inbox.

Perfectly summarized - I highly recommend joining the [Thunderbird Planning](https://thunderbird.topicbox.com/groups/planning) mail list and flagging it! You can also quote me as I regularly get requests for running filters outside of inbox and filters not working. Be aware though that there was a big framework for testing the filters introduced by @[aceman](https://people.mozilla.org/p/aceman) (not sure if he has a github account) - he would probably be the best person to pull on board.

But I have a bade feeling that the wiki needs an update on the [Thunderbird Core Team](https://wiki.mozilla.org/Thunderbird/Core_Team) page, he is listed there but doesn't exist on mozillians.org.

**Comment by RealRaven2000** [comment]
Another good resource to try is [the matrix chat](https://matrix.to/#/#maildev:mozilla.org) of Thunderbird Developers. I am using the [element chat client](https://element.io/) there and get good engagement on the [Thunderbird Add-on Developers](https://matrix.to/#/#tb-addon-developers:mozilla.org) channel. 

Also [Thunderbird Support](https://matrix.to/#/#thunderbird:mozilla.org) may be a good resource

**Comment by ThiloteE** [comment]
I will see what i can do. I just fear i am not qualified enough to hold a conversation on a level and quality that would be required to tackle these issues. I am unable to point towards actual code and say: This is how we can change it. We need this, this and this, it will take this long and then it will be this good, as i lack a background in IT. I am mainly a user. I know how to read and i know some RegEx, but i am not sure that is enough.

The only thing i can say with confidence is: Here are some bug reports. This is a problem.

**Comment by ThiloteE** [comment]
Maybe workaround? @Pilkja Could you report back, if it works for you?

> I also needed to filter based on a folder other than the Inbox, you can use Tags for that. I assume you have filters to move messages from the Inbox to other folders, so you'll have to make another filter before that one to tag the message before the one that moves them (you can have as many tags as you wish i believe) and then another filter between the two that does what you want to messages tagged like that. The end result is that the 1st filter is exactly equal to the 3rd one, but the 1st one only tags while the 3rd one moves, and the 2nd filter does what you want based on messages tagged like that from the 1st filter.
> You will end with messages tagged in the end folder, which I don't know how to remove automatically, but this does work perfectly for me.

- taken from https://forums.mozillazine.org/viewtopic.php?f=39&t=2677919

https://github.com/Betterbird/thunderbird-patches/issues/35 will probably be another workaround, but that one is not implemented yet, as we are speaking.

For the record: A workaround is not the same as a real fix, but maybe a workaround is enough in some use cases 😉

**Comment by Pilkja** [comment]
> @Pilkja Could you report back, if it works for you?

Unfortunately I am on the verge of leaving for Outlook because I need an interactivly working setup with task capabilities, email and note taking (i.e. Outlook and Onenote). I'm not investing any more time in Thunderbird. I'm sorry.

**Comment by RealRaven2000** [comment]
> > @Pilkja Could you report back, if it works for you?
> 
> Unfortunately I am on the verge of leaving for Outlook because I need an interactivly working setup with task capabilities, email and note taking (i.e. Outlook and Onenote). I'm not investing any more time in Thunderbird. I'm sorry.

That's already stuff like this happens all the time. :)  I am leaving the issue open as the discussion might be valuable to some people. By the way there is great work being done on Betterbird now by their team they are finally working on "mixed filter conditions" with a similar UI as apple mail:

https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c67

![attachment](https://user-images.githubusercontent.com/6314674/164219252-4adecdb1-e92a-4dbd-809f-8d353d05078e.png)

**Comment by ThiloteE** [comment]
**Ugly workaround:**

Creaty "dummy account" that acts as intermediate inbox folder for specific filter rules.

This dummy account could be a:

1.    inbox of new e-mail account
 2.   inbox of new local folder
  3.  inbox of new feed (= roughly the same as a local folder)


Workflow:

1. Receive mail in `inbox 1`
2. Run filters (X) on `inbox 1`
   - Filters (X) move or copy mails from `inbox 1` to `Dummy Inbox`
4. Automatically and periodically run filters (Y) on `Dummy inbox`

**Comment by RealRaven2000** [comment]
please check out:

https://github.com/RealRaven2000/FiltaQuilla/issues/318#issuecomment-2636518355

I added the option "run filter on incoming mails" to the folder properties (general tab) - but only for IMAP folders:

![Image](https://github.com/user-attachments/assets/2a53962c-1252-46a8-8075-228bba1dc1a6)

if you can test it there and leave a comment it would be helpful!
